File: Hrdina/src/name/fis/hrdina/generators/WeightedProbBoard.java
package name.fis.hrdina.generators;

import name.fis.hrdina.Alphabet;

public class WeightedProbBoard implements IBoardGenerator
{
	private final Alphabet m_Alphabet;
	
	public WeightedProbBoard (Alphabet alphabet)
	{
		m_Alphabet = alphabet;
	}
	
	@Override
	public char[] GenerateBoard(int edgeSize) {
		int boardSize = edgeSize * edgeSize;
		char[] result = new char[boardSize];
		for (int i = 0; i < boardSize; i++)
		{
			result[i] = m_Alphabet.GetRandomLetterWeighted();
		}
		return result;
	}
	
}

File: pyBoardGen/WeightedProbBoard.py
from GameBoard import GameBoard

class WeightedProbBoard:
	@staticmethod
	def GenerateBoard(edgeSize, alphabet, wordTree, digraphs):
		boardSize = edgeSize * edgeSize
		result = []
		lettersUsed = {}
		for i in range(boardSize):
			letter = alphabet.GetRandomLetterWeighted()

			# prevent any letter from being used more than twice on a board to improve diversity
			while letter in lettersUsed and lettersUsed[letter] >= 2:
				letter = alphabet.GetRandomLetterWeighted()

			if letter in lettersUsed:
				lettersUsed[letter] += 1
			else:
				lettersUsed[letter] = 1
			result += [letter]

		return GameBoard(result, alphabet, wordTree)

